commit ff9aba206a30288f456dfc91584a52ba9927b438
Author: Thomas Loimer <thomas.loimer@tuwien.ac.at>
Date:   Tue Jan 21 20:50:15 2025 +0100

    Allow an arc-box with zero radius, ticket #184
    
    In the pict2e output, a rectangle with rounded corners,
    dashed line type and zero corner-radius would cause a crash.
    Convert rectangles with rounded corners and zero corner-radius to
    regular rectangles.

diff --git a/fig2dev/read.c b/fig2dev/read.c
index 387527b..f643f7e 100644
--- a/fig2dev/read.c
+++ b/fig2dev/read.c
@@ -1001,133 +1001,141 @@ static int
 sanitize_lineobject(F_line *l, int line_no)
 {
 	/* CHANGE if object definitions change! Also, see below. */
 	const char    *obj_name[] = {
 		"rectangle",
 		"polygon",
 		"rectangle with rounded corners",
 		"picture box"
 	};
 
 	if ((l->type == T_BOX || l->type == T_POLYGON ||
 			l->type == T_ARC_BOX || l->type == T_PIC_BOX)) {
 
 		/* close an open box or polygon; testsuite -k open,read.c */
 		if (l->points->x != l->last[0].x ||
 				l->points->y != l->last[0].y) {
 			F_point	*q;
 
 			/* CHANGE THIS if object definitions change! */
 			put_msg("An open %s at line %d - close it.",
 					obj_name[l->type-2], line_no);
 			if (NULL == (Point_malloc(q))) {
 				put_msg(Err_mem);
 				return -1;
 			}
 			/* l->last[] only contains the position, not the link
 			   to the next point; Insert the last point in front. */
 			q->next = l->points;
 			q->x = l->last[0].x;
 			q->y = l->last[0].y;
 			l->points = q;
 			++l->num_points;
 		}
 
 		/* reject incorrect arc-boxes and picture boxes */
 		if (l->type == T_ARC_BOX || l->type == T_PIC_BOX) {
 			if (l->num_points != 5) {
 				/* tests/testsuite -k malformed,read.c */
 				put_msg("A malformed %s at line %d - it has %d "
 						"corners.",
 						obj_name[l->type - 2],
 						line_no, l->num_points - 1);
 				return -1;
 			} else {
 #define IS_RECTANGLE(p, q)	/* p, q is the first and third point */	\
 		((p->x == p->next->x && q->x == q->next->x &&		\
 		  p->y == q->next->y && p->next->y == q->y) ||		\
 		 (p->y == p->next->y && q->y == q->next->y &&		\
 		  p->x == q->next->x && p->next->x == q->x))
 
 				/* tests/testsuite -k distorted,read.c */
 				F_point	*q = l->points->next->next;
 
 				if (!IS_RECTANGLE(l->points, q)) {
 					put_msg("A distorted or inclined %s at "
 							"line %d.",
 							obj_name[l->type - 2],
 							line_no);
 					return -1;
 				}
+				if (l->type == T_ARC_BOX && l->radius == 0) {
+					put_msg("A %s, but zero corner radius "
+							"at line %d - convert "
+							"to a rectangle.",
+							obj_name[l->type - 2],
+							line_no);
+					l->type = T_BOX;
+				}
 			}
 
 		} else { /* T_BOX || T_POLYGON */
 
 			/* too few points - convert to a line */
 			if (l->num_points < 4) {
 				put_msg("A %s with %d points at line %d - "
 						"convert to a polyline.",
 						obj_name[l->type - 2],
 						l->num_points, line_no);
 				l->type = T_POLYLINE;
 				/* the polygon is closed or was closed above,
 				   hence it has at least two points */
 				if (l->num_points == 2) {
 					free(l->points->next);
 					l->points->next = NULL;
 					l->last[0].x = l->points->x;
 					l->last[0].y = l->points->y;
 				} else if (l->num_points == 3) {
 					free(l->points->next->next);
 					l->points->next->next = NULL;
 					l->last[0].x = l->points->next->x;
 					l->last[0].y = l->points->next->y;
 					l->last[1].x = l->points->x;
 					l->last[1].y = l->points->y;
 				} /* else (l->num_points == 1) */
 
 			/* convert misformed rectangles to polygons */
 			} else if (l->type == T_BOX) {
 				F_point	*q = l->points->next->next;
 
 				if (l->num_points != 5) {
 					put_msg("A rectangle with %d corners at"
 							" line %d - convert to "
 							"a polygon.",
 							l->num_points - 1,
 							line_no);
 					l->type = T_POLYGON;
 
 				} else if (!IS_RECTANGLE(l->points, q)) {
 					put_msg("A distorted retangle at line "
 							"%d - convert to a "
 							"polygon.", line_no);
 					l->type = T_POLYGON;
 #undef IS_RECTANGLE
 				}
 			}
 		} /* end l->type == T_BOX || l->type == T_POLYGON */
 
 	} /* l->type == any box or polygon */
 
 	/* above, a closed line might have been converted to a polyline */
 	if (l->type == T_POLYLINE && (l->points->next == NULL ||
 					 l->points->next->next == NULL)) {
 		l->fill_style = UNFILLED;
 		if (l->points->next == NULL) {
 			if (l->for_arrow) {
 				/* tests/testsuite -k polyline,read.c */
 				put_msg("A single point with a forward arrow - "
 						"remove the arrow.");
 				free(l->for_arrow);
 				l->for_arrow = NULL;
 			}
 			if (l->back_arrow) {
 				put_msg("A single point with a backward arrow -"
 						" remove the arrow.");
 				free(l->back_arrow);
 				l->back_arrow = NULL;
 			}
 		}
 	}
 	return 0;
 }
diff --git a/fig2dev/tests/read.at b/fig2dev/tests/read.at
index 588424d..1b4baea 100644
--- a/fig2dev/tests/read.at
+++ b/fig2dev/tests/read.at
@@ -123,6 +123,17 @@ A rectangle with 3 corners at line 11 - convert to a polygon.
 ])
 AT_CLEANUP
 
+AT_SETUP([convert an arc-box with zero radius to a box])
+AT_KEYWORDS(read.c arc-box)
+AT_CHECK([fig2dev -L pict2e <<EOF
+FIG_FILE_TOP
+2 4 1 1 0 0 50 -1 -1 4.0 0 0 0 0 0 5
+	0 0 300 0 300 300 0 300 0 0
+EOF
+],0,ignore,[A rectangle with rounded corners, but zero corner radius at line 11 - convert to a rectangle.
+])
+AT_CLEANUP
+
 AT_SETUP([fail on a malformed arc-box])
 AT_KEYWORDS(read.c malformed arc-box)
 AT_CHECK([fig2dev -L pict2e <<EOF
